### PR TITLE
LeaveGroup and CancelApplication use cases

### LINKED ISSUE #3: [Feature 3] Leave Group or Cancel Application
A feature that lets a user leave a group they are part of, or cancel their application for a group.

### PR DESCRIPTION
**Pull request by aaryavatsa**
Nov. 17:

- Removed some test cases in my interactors (checking whether current user exists is useless!) and by extension, the DsGateways
- Added userInGroup() and groupInUser() methods in DsGateways (for edge cases and testing)
- * RENAMED Presenter interface -> OutputBoundary interface and ResponseFormatter class -> Presenter class (Paul's example has the files named incorrectly, he clarified this in lec and it would be good for us to rename our files to the correct name) *
- Added some methods in the entity classes regarding my use cases
- Fixed some warnings for Leo and Sohee's interfaces (we don't need the keyword public)
- Added Javadocs and method comments

- Presenter class is currently under construction
- I lost my testing code :( git stash fail. I'll rewrite that and have that up soon

### CODE REVIEW

**Comment by aaryavatsa** [comment]
Also y'all how detailed do we have to be in our documentation.

File: src/main/java/LeaveGroup/LeaveGroupResponseModel.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package LeaveGroup;
```
Comment (SharonHsieh22) [inline src/main/java/LeaveGroup/LeaveGroupResponseModel.java:1]: I don't think the response model can interact with the entities?

Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupResponseModel.java:1]: omg yeah you're right. Oops, I misunderstood the response model, I think. I've removed the entitites and am using Strings now. Thank you!

**Review by ipekakyol — DISMISSED** [review]
The use case seems to work great. The previous comments on the response model made me realize that I need to update my response model as well. I know we also discussed using an interface to combine the response models that only return strings. Otherwise, everything looks good!

File: src/main/java/Entities/Group.java [hunk 2]
```diff
@@ -97,5 +97,13 @@ public boolean validateCourseCode(String courseCode){
...
         this.groupName = name;
     }
+    public void removeMember(String username) {
+        this.groupMembers.remove(username);
+    }
+    public void removeApplication(String username) {
+        this.memberRequests.remove(username);
+    }
+
+
```
Comment (soheegoo) [inline src/main/java/Entities/Group.java:107]: Just wondering why these entities are part of this pull request? Have they been changed? Since they are part of mine and Leo's use case i dont think they should be changed in this pull request. Just dont want to have any merge conflicts.

File: src/main/java/CancelApplication/CancelApplicationDsGateway.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import Entities.Group;
+import Entities.User;
+
+public interface CancelApplicationDsGateway {
+    boolean groupExists(String groupname);
+    User getUser(String username);
+    Group getGroup(String groupname);
+
+    /**
+     * Updates the current user after user's applications list has changed. Saves to repository.
```
Comment (helenaglow) [inline src/main/java/CancelApplication/CancelApplicationDsGateway.java:12]: Good use of Java docs!!! They make the files much easier to follow.

File: src/main/java/CancelApplication/CancelApplicationInputBoundary.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package CancelApplication;
+
+public interface CancelApplicationInputBoundary {
+    CancelApplicationResponseModel cancelApplication(CancelApplicationRequestModel requestModel);
```
Comment (helenaglow) [inline src/main/java/CancelApplication/CancelApplicationInputBoundary.java:4]: I'm glad you caught that we should do a more specific method name rather than "create". cancelApplication is a good name as it is specific to your use case.

File: src/main/java/CancelApplication/CancelApplicationDsGateway.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package CancelApplication;
+
+import Entities.Group;
+import Entities.User;
+
+public interface CancelApplicationDsGateway {
+    boolean groupExists(String groupname);
+    User getUser(String username);
+    Group getGroup(String groupname);
+
```
Comment (helenaglow) [inline src/main/java/CancelApplication/CancelApplicationDsGateway.java:10]: I think one thing that we should talk about is what we should name our DsGateway methods. Of course, this gate way is specific to you, but we all use the getUser call. I personally have getUserByName(). So, it would be best to discuss what we should name shared method calls to avoid confusion.

File: src/main/java/CancelApplication/CancelApplicationInteractor.java [hunk 6]
```diff
@@ -3,17 +3,32 @@
...
         this.dsGateway = dsGateway;
         this.outputBoundary = outputBoundary;
     }
 
+    /**
+     * @param requestModel the requestModel for the cancelApplication use case
+     * @return a successView or failView depending on whether the user has cancelled their group application
+     */
     @Override
     public CancelApplicationResponseModel cancelApplication(CancelApplicationRequestModel requestModel) {
```
Comment (hannahpan1) [inline src/main/java/CancelApplication/CancelApplicationInteractor.java:34]: Looks good! You covered all major cases and everything is really easy to follow. I think it might also be helpful to check whether user exists too before checking whether user is in the group because User user = dsGateway.getUser(requestModel.getUsername()); might return null even though it shouldn't if the user name is displayed incorrectly.

File: src/main/java/CancelApplication/CancelApplicationPresenter.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package CancelApplication;
+
+public class CancelApplicationPresenter implements CancelApplicationOutputBoundary {
```
Comment (helenaglow) [inline src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: Is having both an output boundary and Presenter necessary? Why have these if they return null?

From my understanding Paul's presenter is his output boundary.

File: src/main/java/LeaveGroup/LeaveGroupInteractor.java [hunk 8]
```diff
@@ -5,18 +5,33 @@
...
 
         this.dsGateway = dsGateway;
         this.outputBoundary = outputBoundary;
     }
 
+    /**
+     * @param requestModel the requestModel for the leaveGroup use case
+     * @return a successView or failView depending on whether the user has left the group
+     */
     @Override
```
Comment (hannahpan1) [inline src/main/java/LeaveGroup/LeaveGroupInteractor.java:35]: Again here it might be helpful to have a check that user exists.

Comment (hannahpan1) [reply src/main/java/LeaveGroup/LeaveGroupInteractor.java:35]: Also I think it would be nice to have a more informative message for the user if they are the group leader. Something like "User is the group leader. Group leaders must be reassigned to another user before leaving the group."

**Review by hannahpan1 — APPROVED** [review]

File: src/main/java/LeaveGroup/LeaveGroupDsGateway.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    void updateGroup(String groupname);
+    boolean userInGroup(String username, String groupName);
+
+    /**
+     * @param username the username of the current user
+     * @param groupName the name of the group
+     * @return whether the user's members list contains the group's name.
+     */
+    boolean groupInUser(String username, String groupName);
+}
```
Comment (helenaglow) [inline src/main/java/LeaveGroup/LeaveGroupDsGateway.java:30]: This will be fixed when we make one general gateway, but for you own use case have you considered combining your two gateways? they share similar functions such as getUser, groupExists, and so on. 

Also, for updateGroup(String groupname) how is the repo going to know how to update the group (i.e. how will it know that the users have changed)? especially if it just has group name. Maybe I am missing something.

(Update: we are not doing a general
Ds gateway it's a clear violation of clean architecture)

File: src/main/java/LeaveGroup/LeaveGroupInteractor.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+     * @return a successView or failView depending on whether the user has left the group
+     */
+    @Override
+    public LeaveGroupResponseModel leaveGroup(LeaveGroupRequestModel requestModel) {
+
+        if (!dsGateway.groupExists(requestModel.getGroupname())) {
+            return outputBoundary.prepareFailureView("Group does not exist.");
+        }
+
+        User user = dsGateway.getUser(requestModel.getUsername());
```
Comment (helenaglow) [inline src/main/java/LeaveGroup/LeaveGroupInteractor.java:42]: I like how you included descriptive error messages for each error, which you have covered for most cases. It will be easier for the user to know what they did wrong, and/or when you are testing the code.

File: src/main/java/LeaveGroup/LeaveGroupPresenter.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package LeaveGroup;
+
+public class LeaveGroupPresenter implements LeaveGroupOutputBoundary {
```
Comment (helenaglow) [inline src/main/java/LeaveGroup/LeaveGroupPresenter.java:3]: Again, what is the purpose of the presenter?

File: src/main/java/CancelApplication/CancelApplicationDsGateway.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Updates the current user after user's applications list has changed. Saves to repository.
+     * @param username the username of the current user
+     */
+    void updateUser(String username);
```
Comment (helenaglow) [inline src/main/java/CancelApplication/CancelApplicationDsGateway.java:15]: I mentioned this for your other use case, but how will we know what to update in the User? I am a little confused

**Review by helenaglow — APPROVED** [review]
I have left comments! Overall, you followed clean architecture very well! The java docs were very helpful when understanding your code. 

My main comments are on the presenter and output boundary, and the ds gateways.

File: src/main/java/CancelApplication/CancelApplicationDsGateway.java (hunk 5 above)
Comment (aaryavatsa) [reply src/main/java/CancelApplication/CancelApplicationDsGateway.java:10]: Yeah, that's a great point. Once we set up the repository and have it implement all our gateways -- I think it'll be very clear to see any repetitions. Of course, we could also discuss this beforehand.

File: src/main/java/CancelApplication/CancelApplicationPresenter.java (hunk 7 above)
Comment (aaryavatsa) [reply src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: Oops, should have left a comment on my code for this. Yes, you're correct that Paul's presenter is actually his output boundary.

His ACTUAL presenter is the file named "ResponseFormatter" in the screens package. So in my code, I haven't actually written anything for my presenter because I'm not sure yet what info the presenter will pass along to the screens. It's returning null (for now) because my use case is unfinished until I figure out what I want my screen to display.

File: src/main/java/CancelApplication/CancelApplicationInteractor.java (hunk 6 above)
Comment (aaryavatsa) [reply src/main/java/CancelApplication/CancelApplicationInteractor.java:34]: See, I was unsure about that. Originally I had that check, as you said, because the username might be incorrect which would raise a nullpointer exception since I can't get the user from the repo.
But now we have a current user. It's initialized as soon as you log in (from my understanding after talking to Leo) and so his usecase sets the current user, which is the one currently logged in, which means the current user HAS to exist because they've already logged in. 
So I don't know if having this check is redundant. I'll talk to Sharon and Ipek about this -- I think they have the same check.

File: src/main/java/LeaveGroup/LeaveGroupInteractor.java (hunk 8 above)
Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupInteractor.java:35]: > Also I think it would be nice to have a more informative message for the user if they are the group leader. Something like "User is the group leader. Group leaders must be reassigned to another user before leaving the group."

Okay but do I have this specific error message in the use case or is it the Presenter's job to output a more specific error message that can be shown to the user? I wasn't sure. I thought maybe the presenter could do that, as the user never sees this interactor class and is only seeing a nicely formatted message from the presenter. But I'm not sure. I wanted to make the method concise.

File: src/main/java/LeaveGroup/LeaveGroupDsGateway.java (hunk 9 above)
Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupDsGateway.java:30]: My two dsGateways are the exact same, and I think they're pretty identical to Sharon and Ipek's. We talked to Paul about this issue -- he didn't really answer the question about HOW we could combine gateways across usecases, but he said that it wasn't an issue for different use cases to have similar methods in their gateways since each use case should be totally independent from the rest. He said that as long as the repo implements all the gateways, there is no problem.

I'm not totally convinced by this answer, and I think I'll follow up with him again.

Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupDsGateway.java:30]: > Also, for updateGroup(String groupname) how is the repo going to know how to update the group (i.e. how will it know that the users have changed)? especially if it just has group name. Maybe I am missing something.

Oh you're so right actually thank you for catching this! I forgot to update this method after we figured out the current user thing.. the method should have the current user username and the groupname as parameters.

The repository has the ability to find the group object from its groupname (because it contains all Groups). So the updateGroup() method implemented by the repo will be able to get the Group object from the groupname and then mutate its members list and then call the save() method in the repo to record this change. Does this make sense, feel free to correct me if I'm not understanding the repo correctly.

File: src/main/java/LeaveGroup/LeaveGroupPresenter.java (hunk 11 above)
Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupPresenter.java:3]: review comment for presenter above ^

File: src/main/java/CancelApplication/CancelApplicationDsGateway.java (hunk 12 above)
Comment (aaryavatsa) [reply src/main/java/CancelApplication/CancelApplicationDsGateway.java:15]: same as other usecase, thank you for catching!

File: src/main/java/CancelApplication/CancelApplicationPresenter.java (hunk 7 above)
Comment (helenaglow) [reply src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: Oh makes sense! Thank you! Is there a difference between having the presenter in the use case package rather than a screen? 

I see your comment earlier. I must have forgotten it when I was reading your code. From my understanding, your implementation would just remove the group from the list, correct me if I am wrong. So maybe return a list of groups without the one they removed (as strings of course) or a message that the group was removed. This would largely depend on how we display the home screen. I really liked your design on our blueprint, it is very clean and clear. If we did it that way, I would think that the groups have already been somehow presented and you are just refreshing the screen. I hope that makes sense. We have not really discussed home screen as we were really focused on our own use cases, but I think it should be useful when implementing your presenter :)

File: src/main/java/CancelApplication/CancelApplicationInteractor.java (hunk 6 above)
Comment (helenaglow) [reply src/main/java/CancelApplication/CancelApplicationInteractor.java:34]: I agree with Aarya. It does seem rather redundant. However, for us not to need that check, we would have to be 100% certain that we are grabbing the username correctly, and/or the user creation use case functions correctly. If we work under that assumption (I think we should otherwise we have a faulty app which is a problem in it of itself), then I do not believe it needs to be counted. Of course, check with Ipek and Sharon as they are doing something very similar.

File: src/main/java/LeaveGroup/LeaveGroupInteractor.java (hunk 8 above)
Comment (helenaglow) [reply src/main/java/LeaveGroup/LeaveGroupInteractor.java:35]: That is actually a very interesting point. I am pretty sure the messages are in the interactor (from my understanding) and you just send your results to the presenter (through the output boundary).  How would you have the presenter know what error message to present?

File: src/main/java/CancelApplication/CancelApplicationPresenter.java (hunk 7 above)
Comment (aaryavatsa) [reply src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: > Oh makes sense! Thank you! Is there a difference between having the presenter in the use case package rather than a screen?

Oops, you're right! Forgot to move this to the screens package after I renamed it. Both the Presenter and the Controller should be in the screens package, yeah.

> So maybe return a list of groups without the one they removed (as strings of course) or a message that the group was removed. This would largely depend on how we display the home screen.

Yes, I totally agree. After we figure out the formatting for the home screen and the other "linking" screens that transition between usecases -- I'll definitely have to modify my presenter to make sure the correct info is given to the screen for the other usecases or whoever. I think this is fine for the time being though, at least until we sit and discuss the home screen as a group?

File: src/main/java/LeaveGroup/LeaveGroupDsGateway.java (hunk 9 above)
Comment (helenaglow) [reply src/main/java/LeaveGroup/LeaveGroupDsGateway.java:30]: > My two dsGateways are the exact same, and I think they're pretty identical to Sharon and Ipek's. We talked to Paul about this issue -- he didn't really answer the question about HOW we could combine gateways across usecases, but he said that it wasn't an issue for different use cases to have similar methods in their gateways since each use case should be totally independent from the rest. He said that as long as the repo implements all the gateways, there is no problem.
> 
> I'm not totally convinced by this answer, and I think I'll follow up with him again.

That's actually so interesting hmmm. I guess it would make sense to have each use case have its own dsGateway otherwise our application would be way too dependent on our General DsGateway. If we changed something it might cause errors in all use cases rather than just one. Okay that makes sense thank you for the clarification!

File: src/main/java/LeaveGroup/LeaveGroupInteractor.java (hunk 8 above)
Comment (aaryavatsa) [reply src/main/java/LeaveGroup/LeaveGroupInteractor.java:35]: Hmm, alright. I was thinking that since the Presenter is already supposed to format the given string to make it more appropriate for the screen, why can't it modify the "User left group" message into something more descriptive as well? But now that I'm typing this out I think it would make more sense just to have the correct message in the interactor to begin with. I'll change that. Thanks Hannah and Helena!

File: src/main/java/LeaveGroup/LeaveGroupDsGateway.java (hunk 9 above)
Comment (helenaglow) [reply src/main/java/LeaveGroup/LeaveGroupDsGateway.java:30]: > 

Great that makes sense! Good implementation idea! Having the username as the input would make it easier to remove it from the group's member list, or application list. It also seems like the best way to do so while also maintaining clear architecture and not passing around entities.

File: src/main/java/CancelApplication/CancelApplicationPresenter.java (hunk 7 above)
Comment (helenaglow) [reply src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: Yes totally! We really need to meet up and discuss. I can understand how it can be difficult to decide how to present when we aren't even certain of where or how it will be presented :(. For now, maybe brainstorm and think of all the possibilities🤩.

Comment (ipekakyol) [reply src/main/java/CancelApplication/CancelApplicationPresenter.java:3]: After looking at this discussion, I believe I should also have an output boundary and presenter and will implement it following the changes made here.

File: src/main/java/CancelApplication/CancelApplicationInteractor.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    @Override
+    public CancelApplicationResponseModel cancelApplication(CancelApplicationRequestModel requestModel) {
+
+        if (!dsGateway.groupExists(requestModel.getGroupname())) {
+            return outputBoundary.prepareFailureView("Group does not exist.");
+        }
+
+        User user = dsGateway.getUser(requestModel.getUsername());
+        Group group = dsGateway.getGroup(requestModel.getGroupname());
```
Comment (ipekakyol) [inline src/main/java/CancelApplication/CancelApplicationInteractor.java:40]: One thing that I was wondering is that should we get the user this way from the gateway. I had initialized the user through CurrentUser.getInstance().getUser(), does this approach make sense or should I also change mine?

**Review by ipekakyol — APPROVED** [review]
